Add LIP "Introduce cross-chain messages" (#102)
LiskHQ/lips: proposals/lip-Introduce_cross_chain_messages.md

```
LIP: <LIP number>
Title: Introduce cross-chain messages
Author: Maxime Gagnebin <maxime.gagnebin@lightcurve.io>
Type: Standards Track
Created: <YYYY-MM-DD>
Updated: <YYYY-MM-DD>

> Discussions-To: https://research.lisk.com/t/introduce-cross-chain-messages/299

Reviewer: The above item is not in the order defined in LIP 0001.

Requires: Introduce cross-chain update mechanism
```

## Abstract

This proposal introduces the cross-chain message schema, the generic message processing and the base error handling. 
Defining a base cross-chain message allows all chains in the ecosystem to read and understand the base properties of messages.

> The proposal also introduces three standard messages used by the Interoperability module, the cross-chain update receipt, the channel terminated receipt and the registration message.

Author: removed

## Copyright

This LIP is licensed under the [Creative Commons Zero 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/).

## Motivation

To achieve interoperability, chains need to exchange information by sending messages to each other. 
To this end, we introduce a new message format, which we call a _cross-chain message_ and define in this proposal the base schema of cross-chain messages. 
Specifying a unified base message schema allows all chains in the Lisk ecosystem to deserialize and read cross-chain messages. 

> To achieve some of its basic functionalities, the Interoperability module uses three standard messages. 

Author: removed

## Rationale

Messages in the Lisk ecosystem will be included in multiple chains. 
It is therefore important that all chains can deserialize and process the base part of cross-chain messages in the same way. 
The properties of the message contain all the information necessary for this purpose while trying to keep a minimal size.

### Cross-chain Message Properties

In the following, we list all the properties of a cross-chain message.

#### Sending Chain ID and Receiving Chain ID

Used to identify the chains exchanging the cross-chain message. 
On the mainchain the receiving chain ID is read to route the message to the corresponding chain outbox, as specified in [LIP "Introduce cross-chain update mechanism"][CCU-LIP]. 
The sending chain ID is used for example if the message triggers an error and has to be sent back.

#### Index

When a cross-chain message is created and added to the partner chain outbox, the size of the outbox at that point is added to the message in the `index` property. 
This allows all messages to be distinct and to be tracked throughout the ecosystem.

> 

Author: I reworded slightly, but not really sure what you mean here.

> 

Reviewer: The previous sentence

>  This allows all messages to be distinct. 

did not make it clear whether the message emitted by one chain are distinct or in the Lisk ecosystem, all messages are unique. As we now have message IDs, the sentence on line 68 could be changed to

>  This property ensures that all messages in the Lisk ecosystem are distinct  which is important for identifying messages and tracking them throughout the ecosystem.

> 

Reviewer: Note that the tuple should also be `(sendingChainID, receivingChainID, index)`

> 

Reviewer: All good, can be resolved.

#### Module ID and Cross-chain Command ID

Once the message has reached the recipient chain, the two properties `moduleID` and `crossChainCommandID` specify which logic should be used to validate and execute the message. 
The Interoperability module handles the message in case the required logic is not available on the chain. 
This brings the benefit that sending chains do not need to monitor all other chains in the ecosystem and the modules they support. 
We are following an optimistic approach in which all messages will be delivered and error handling will kick in when it is needed. 

> In that regard, users should be aware that valid messages will usually not trigger a response. 

Author: Yes, this is up to the module. For all the modules we have specified so far, we don't have a return message when the CCM is successful. In that regard, I think it's fair to warn users that valid messages will usually not trigger a response. What do you think should be written differently?

> In that regard, users should be aware that valid messages will usually not trigger a response. 

Reviewer: Once suggestion.

> Once the message has reached the recipient chain, the two properties `moduleID` and `crossChainCommandID` specify which logic should be used to validate and execute the message. 
The Interoperability module perfoms some basic validations for all messages and then forwards the cross-chain message to the respective module which is responsible for the validation and execution. In case the required logic is not available on the chain, the Inoperability module sends a message with an error status back to the recipient chain provided the message included enough fees.
This approach brings the benefit that sending chains do not need to monitor all other chains in the ecosystem and the modules they support. 

> Apart from the basic validations and error handling, the Interoperability module does not intervene in the cross-chain messaging between modules and is agnostic of the message protocol that modules use to communicate with each other. In particular, the messaging protocol between modules can choose whether the reception and successful execution of a message is confirmed with a return message. 
For the Token and NFT module,  we are following an optimistic approach in which we assume that all messages are successfully delivered unless an error message is returned. In this case, users should be aware that valid messages send from a chain will not trigger a response message to the chain.

> In that regard, users should be aware that valid messages will usually not trigger a response. 

Reviewer: The second paragraph could also be in a separate section about error handling.

> In that regard, users should be aware that valid messages will usually not trigger a response. 

Author: modified the `moduleID` paragraph and added some information to the `message tracking` section.

> In that regard, users should be aware that valid messages will usually not trigger a response. 

Reviewer: Sounds good.

The constant table lists the different status codes defined by the Interoperability module. 
To allow for future updates or improvements of the Interoperability module, we restrict other modules from using status codes up to 63 (included). 
Other modules may use other status codes larger or equal to 64.

#### Parameters

The `params` property of the messages is defined by each module and can follow any schema, similar to the `params` property of a transaction. 
The `params` property is not deserialized or validated by the Interoperability module.

In the Lisk ecosystem, all cross-chain messages are routed through the mainchain. 
This means that messages should always have a sufficiently small size in order to be easily included in mainchain blocks. 
As the mainchain payload size limit is 15 KiB, and [other properties][CCU-LIP] in the cross-chain updates will not be larger than 4 KiB, we limit the message size to 10 KiB. 
To guarantee that all messages can be included and handled, sidechains in the Lisk ecosystem should have a payload size limit equal to or greater than 15 KiB (15 x 1024 bytes).

### Message Tracking

> Tracking messages throughout the ecosystem could be a challenge, particularly if the messages are errored (and hence get a new index property from the erroring chain).

Author: Changed the wording.
Added a section on the message ID (in the rationale and in the specs - even though the ID is not really used anywhere in the protocol so far).

When the first cross-chain update is sent to the sidechain, the equality between the properties in the cross-chain message and the ones in the interoperability store is verified.

### Sidechain Terminated Message

The role of the sidechain terminated message is to inform sidechains that another sidechain has been terminated on the mainchain and is unable to receive messages. 
The message contains the ID of the terminated chain as well as the last certified state root of the terminated sidechain (as certified on the mainchain).
This value is used for the creation of the terminated chain account (on the sidechain receiving the CCM), allowing state recoveries.

This message allows to inform sidechains about other terminated sidechains efficiently.
Indeed, this message will automatically trigger the creation of the terminated sidechain account as soon as the first message is unable to be delivered.
This also prevents further messages to be sent to already terminated sidechains.

## Specification

Cross-chain messages (CCM) are used by modules to execute actions on other chains in the ecosystem. 
They are part of the Interoperability module.

### Notation and Constants

> The following constants are used throughout the document, multiple of those constants are shared with the other LIPs defining the Interoperability module and all of the needed constants for the interoperability module are defined in [LIP "Introduce Interoperability module"][base-interoperability-LIP]. 

Reviewer: ```suggestion
The following constants are used throughout the document, multiple of those constants are shared with the other LIPs defining the Interoperability module and all of the needed constants for the Interoperability module are defined in [LIP "Introduce Interoperability module"][base-interoperability-LIP]. 
```

Author: done

        }
    }
}
```

### Internal Functions

#### createCrossChainMessage

The following logic should be used to create new cross chain messages:

```python
createCrossChainMessage(moduleID, crossChainCommandID, receivingChainID, fee, params):

    return message = {
               "index": 0,
               "moduleID": moduleID,
               "crossChainCommandID": crossChainCommandID,
               "sendingChainID": chainID of the current chain,
               "receivingChainID": receivingChainID,
               "fee": fee,
               "status": CCM_STATUS_OK,
               "params": params
           }
```

#### validateFormat 

All cross-chain messages `CCM` must have the correct format, which is checked by the following logic:

```python
validateFormat(CCM):
   if size(CCM) > MAX_CCM_SIZE bytes: 
      return False
   if CCM does not follow crossChainMessageSchema: 
      return False
   return True
```

#### process

When processing a cross-chain message, with a valid format, `CCM`, follow the logic below:

```python
process(CCM):
   let ownChainID be the chainID of the current chain 

   if CCM.receivingChainID == ownChainID:
      tryToExecute(CCM)

>    else: // CCM.receivingChainID != ownChainID

Author: indeed, removed

         newCCM.status = CCM_STATUS_CROSS_CHAIN_COMMAND_NOT_SUPPORTED 
         swap newCCM.receivingChainID and newCCM.sendingChainID
         process(newCCM) 
```

```python
tryToForward(CCM):
   let partnerChainID = getPartnerChainID(CCM.receivingChainID)

   if account(partnerChainID).status == CHAIN_ACTIVE and isLive(partnerChainID):
      addToOutbox(partnerChainID, CCM)
   elif CCM.status == CCM_STATUS_OK:
      newCCM = CCM
      newCCM.fee = 0
      newCCM.status = CCM_STATUS_CHANNEL_UNAVAILABLE  
      swap newCCM.receivingChainID and newCCM.sendingChainID
      process(newCCM)
      
      STMParams = {chainID: CCM.receivingChainID,
                  stateRoot: account(CCM.receivingChainID).lastCertifiedStateRoot}
                  serialized using sidechainTerminatedCCMParamsSchema
      STM = createCrossChainMessage(MODULE_ID_INTEROPERABILITY, 
                                    CROSS_CHAIN_COMMAND_ID_SIDECHAIN_TERMINATED,
                                    CCM.sendingChainID,
                                    0,
                                    STMParams)
      addToOutbox(STM)
   else:
      CCM is discarded and has no further effect
```

> The `addToOutbox` function is defined in [LIP "Introduce Interoperability module"][base-interoperability-LIP].

Author: yes, thanks, added.

### Cross-chain Commands

#### Cross-chain Update Receipt

The cross-chain update receipt (CCU receipt) is a CCM with 

*   `moduleID = MODULE_ID_INTEROPERABILITY`,
*   `crossChainCommandID  = CROSS_CHAIN_COMMAND_ID_CCU_RECEIPT`.

The main role of the cross-chain update receipt is to inform chains that a cross-chain update transaction has been posted on the partner chain.

##### Parameters

The `params` schema for CCU receipts is:

```java
CCUReceiptParamsSchema = {
    "type": "object",
    "required": [
        "paidFee", 
        "relayerPublicKey", 
        "partnerChainInboxSize"
    ],
    "properties": {
        "paidFee": {
            "dataType": "uint64",
            "fieldNumber": 1 
        },
        "relayerPublicKey": {
            "dataType": "bytes",
            "fieldNumber": 2 
        },
        "partnerChainInboxSize": {
            "dataType": "uint64",
            "fieldNumber": 3
        }
    }
}
```
*   `paidFee`: fee paid by the relayer for the cross-chain update transaction,

> *   `relayerPublicKey`: public key of the relayer posting the cross-chain update transaction,

Author: changed as discussed in our standup

*   `partnerChainInboxSize`: size of the partner chain inbox.

##### Creating Cross-chain Update Receipts

A CCU receipt is created by the Interoperability module upon acting on a cross-chain update transaction as specified in [LIP "Introduce cross-chain update mechanism"][CCU-LIP].

##### Executing Cross-chain Update Receipts

> When a CCU receipt `CCUR` is received, the following is done:

Reviewer: Our standard is normally camelCase for variables and function, upper snake case for constants. Using upper case also messes up the syntax highlighting.

Feel free to ignore this one though as the LIP has been already reviewed.

Author: changed

#### Channel Terminated Message 

The channel terminated message is a CCM with 

*   `moduleID = MODULE_ID_INTEROPERABILITY`,
*   `crossChainCommandID  = CROSS_CHAIN_COMMAND_ID_CHANNEL_TERMINATED`.

The role of the channel terminated message is to inform chains that their channel has been terminated on the mainchain.

##### Parameters

The `params` schema for channel terminated is:

```java
channelTerminatedCCMParamsSchema = {
    "type": "object",
    "required" : ["partnerChainInboxSize"],
    "properties": {
        "partnerChainInboxSize": {
            "dataType": "uint64",
            "fieldNumber": 1
        }
    }
}
```
*  `partnerChainInboxSize`: size of the partner chain inbox.

##### Creating Channel Terminated Messages

A channel terminated message is created by the Interoperability module when [terminating a chain account][base-interoperability-LIP-terminationFunction], for example when encountering a malicious cross-chain update transaction or malicious cross-chain message.

##### Executing Channel Terminated Messages

When a channel terminated message `CTM` is received, the following is done: 

> ```

Author: added

account(CTM.sendingChainID).status = CHAIN_TERMINATED

if account(CTM.sendingChainID).partnerChainInboxSize < CTM.params.partnerChainInboxSize 
    set account(CTM.sendingChainID).partnerChainInboxSize = CTM.params.partnerChainInboxSize
```

#### Registration Message 

The registration message is a CCM with 

*   `moduleID = MODULE_ID_INTEROPERABILITY`,
*   `crossChainCommandID  = CROSS_CHAIN_COMMAND_ID_REGISTRATION`.

The role of the registration message is to guarantee that a channel was opened on the sidechain with the correct chain ID and network ID.

##### Parameters

The `params` schema for the registration CCM is:

```java
registrationCCMParamsSchema = {
    "type": "object",

>     "required" : ["networkID", "name"]

Reviewer: ```suggestion
    "required" : ["networkID", "name"],
```

Author: don

##### Executing Registration Message

When a registration message `RM` is executed, the following is done: 

```python
let ownName and ownChainID be the name and ID properties (respectively) of the deserialized value of account(0) 

if CCM.index != 0 
or ownChainID != CCM.receivingChainID
or ownName != CCM.params.name
or CCM.params.networkID does not equal the chain's networkID:
    terminateChain(RM.sendingChainID)
```
The `terminateChain` function is defined in [LIP "Introduce Interoperability module"][base-interoperability-LIP].

#### Sidechain Terminated Message 

The sidechain terminated message is a CCM with 

*   `moduleID = MODULE_ID_INTEROPERABILITY`.
*   `crossChainCommandID  = CROSS_CHAIN_COMMAND_ID_SIDECHAIN_TERMINATED`.

The role of the sidechain terminated message is to inform sidechains that another sidechain is unable to receive messages. 
This then triggers the creation of the terminated chain account, allowing state recoveries.

##### Parameters

The `params` schema for the sidechain terminated CCM is:

```java
sidechainTerminatedCCMParamsSchema = {
    "type": "object",

>     "required" : ["chainID", "stateRoot"]

Author: done

## Rationale

Messages in the Lisk ecosystem will be included in multiple chains. 
It is therefore important that all chains can deserialize and process the base part of cross-chain messages in the same way. 
The properties of the message contain all the information necessary for this purpose while trying to keep a minimal size.

### Cross-chain Message Properties

In the following, we list all the properties of a cross-chain message.

#### Sending Chain ID and Receiving Chain ID

Used to identify the chains exchanging the cross-chain message. 
On the mainchain, the receiving chain ID is read to route the message to the corresponding chain outbox, as specified in [LIP "Introduce cross-chain update mechanism"][CCU-LIP]. 
The sending chain ID is used, for example, if the message triggers an error and has to be sent back.

#### Index

When a cross-chain message is created and added to the partner chain outbox, the size of the outbox at that point is added to the message in the `index` property. 

> This allows all messages to be uniquely identified by the `(sendingChain, receivingChain, index)` tuple. 

Reviewer: ```suggestion
This allows all messages to be uniquely identified by the `(sendingChainID, receivingChainID, index)` tuple. 
```

> This property is important to identify messages and to track them throughout the ecosystem.

Reviewer: ```suggestion
The uniqueness of messages in the Lisk ecosystem is important for identifying messages with message IDs and tracking them throughout the ecosystem.
```

Reviewer: As we use the message ID for tracking not the index itself.

Author: added

The constant table lists the different status codes defined by the Interoperability module. 
To allow for future updates or improvements of the Interoperability module, we restrict other modules from using status codes up to 63 (included). 
Other modules may use other status codes larger or equal to 64.

#### Parameters

The `params` property of the messages is defined by each module and can follow any schema, similar to the `params` property of a transaction. 
The `params` property is not deserialized or validated by the Interoperability module.

In the Lisk ecosystem, all cross-chain messages are routed through the mainchain. 
This means that messages should always have a sufficiently small size in order to be easily included in mainchain blocks. 
As the mainchain payload size limit is 15 KiB, and [other properties][CCU-LIP] in the cross-chain updates will not be larger than 4 KiB, we limit the message size to 10 KiB. 
To guarantee that all messages can be included and handled, sidechains in the Lisk ecosystem should have a payload size limit equal to or greater than 15 KiB (15 x 1024 bytes).

### Message ID

> As all messages emitted in the Lisk ecoystem are distinct (thanks to the `index` property), we can assign a unique ID to them.

Reviewer 2: Don't you also need the sending chain to completely identify it? Maybe we could add that: "All messages in the Lisk ecosystem are uniquely identified by their `index` and `sendingChain` properties." or something similar

Author: modified the sentence and added something similar to your suggestion to the `Index` rationale.

> As all messages emitted in the Lisk ecoystem are distinct (thanks to the unicity of the `(sendingChain, receivingChain, index)` trio), we can assign a unique ID to them.

Reviewer: ```suggestion
As all messages emitted in the Lisk ecoystem are distinct (thanks to the unicity of the `(sendingChainID, receivingChainID, index)` trio), we can assign a unique ID to them.
```

The message ID is obtained in the same way as other IDs in the protocol, namely by hashing the serialized object.
This assigns to each message a 32-byte value that can be used whenever the message needs to be referenced.

### Message Tracking

Tracking messages throughout the ecosystem could be a challenge, particularly if the messages are errored (and hence spawn new messages from the erroring chain).
However, this can easily be done using the logs emitted by the CCM processing. 
Indeed, messages which are errored and which send a response message to their sending chain, will log the ID of the return message,

> this allows UI products to display the current status of CCMs (or the status of the returned message if needed).

Reviewer 2: ```suggestion
allowing off-chain services to display the current status of CCMs (or the status of the returned message if needed).
```

Reviewer 2: To reuse the language we introduce in the module LIPs (and probably in the chain architecture LIP)

Author: done

```
The `addToOutbox`, `getPartnerChainID` and `isLive` functions are defined in [LIP "Introduce Interoperability module"][base-interoperability-LIP].

### Cross-chain Commands

#### Cross-chain Update Receipt

The cross-chain update receipt (CCU receipt) is a CCM with 

*   `moduleID = MODULE_ID_INTEROPERABILITY`,
*   `crossChainCommandID  = CROSS_CHAIN_COMMAND_ID_CCU_RECEIPT`.

The main role of the cross-chain update receipt is to inform chains that a cross-chain update transaction has been posted on the partner chain.

##### Parameters

The `params` schema for CCU receipts is:

```java
ccuReceiptParamsSchema = {
    "type": "object",
    "required": [
        "paidFee", 
        "relayerPublicKey", 
        "partnerChainInboxSize"
    ],
    "properties": {
        "paidFee": {
            "dataType": "uint64",
            "fieldNumber": 1 
        },
        "relayerAddress": {
            "dataType": "bytes",
            "fieldNumber": 2 
        },
        "partnerChainInboxSize": {
            "dataType": "uint64",
            "fieldNumber": 3
        }
    }
}
```
*   `paidFee`: fee paid by the relayer for the cross-chain update transaction,

> *   `relayerAddress`: address of the relayer posting the cross-chain update transaction. This property must be 20 bytes long.

Reviewer 2: ```suggestion
*   `relayerAddress`: address of the relayer posting the cross-chain update transaction. This property must be 20 bytes long,
```

Reviewer 2: or end with a period the previous bullet point, for consistency

Author: ja, ended the previous one with a bullet.

To this end, we introduce a new message format, which we call a _cross-chain message_ and define in this proposal the base schema of cross-chain messages. 
Specifying a unified base message schema allows all chains in the Lisk ecosystem to deserialize and read cross-chain messages. 

To achieve some of its basic functionalities, the Interoperability module uses four messages. 
The first one, the _cross-chain update receipt_, is sent back to the partner chain (the chain with which the current chain is exchanging messages) whenever a cross-chain update transaction from this chain gets included. 
The cross-chain update receipt serves to update a few properties of the interoperability store, and in general to acknowledge the inclusion of cross-chain update transactions. 
The second one, the _channel terminated message_, is sent to chains which have been terminated. 
This message also updates a few properties of the interoperability store to keep them synchronized between chains. 
The third one, the _registration message_, is used when registering a chain. 
This message activates the channel between the chains.

> Furthermore, it serves as a guarantee that the correct chain ID and network ID were used when the registration transaction was sent on the sidechain.

Reviewer 2: ```suggestion
Furthermore, it serves as a guarantee that the correct name and network ID were used when the registration transaction was sent on the sidechain.
```

Reviewer: > registration transaction was sent on the sidechain.
Shouldn't it also be "was sent on the other chain."

Whenever a cross-chain update transaction is posted on the mainchain, the poster gets compensation from the pool.

#### Channel Terminated Message

The role of the channel terminated message is to inform chains that their channel has been terminated on the mainchain. 
The chain receiving this message can then also close the channel to the mainchain. 
This is helpful in preventing users from sending transactions to a chain whilst the cross-chain update transaction will be invalid. 
Note that the Interoperability module is designed in such a way that no channel should be terminated while the sidechain is respecting the Lisk interoperability protocol. 
Sending and receiving this message should therefore be a rare occurrence.

The termination message contains the last inbox size of the sidechain. 
This allows the sidechain to know exactly which messages were executed on the mainchain and which are still pending at the time of termination.

#### Registration Message

The role of the registration message is to change the sending chain status from `CHAIN_REGISTERED` to `CHAIN_ACTIVE`. 

> Further, it guarantees that a channel was opened on the sidechain with the correct chain ID and network ID. 

Reviewer 2: ```suggestion
Further, it guarantees that a channel was opened on the sidechain with the correct name and network ID. 
```

Reviewer 2: In reality you also explain this in the following sentences, so you could just drop this sentence

Author: removed the sentence

Reviewer: Looking at the code, aren't all 3 values name, chain ID and network ID validated?

```

> ##### Executing Registration Message

Reviewer: Don't we need to activate the channel here?

Author: The activation is actually done by the CCU execution. I updated the motivation and rationale to be more clear on this point.
